feat(aten::to): Add support for cast layer conversion

peri044 opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes issues mentioned in [https://github.com/NVIDIA/TRTorch/discussions/358](url)

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

narendasan commented on core/conversion/converters/impl/cast.cpp:
```diff
@@ -0,0 +1,51 @@
+               return register_cast_layer(ctx, n, self, output_dtype);
+             }})
+        .pattern(
+            {"aten::to.device(Tensor self, Device device, int dtype, bool non_blocking=False, bool copy=False, int? memory_format=None) -> (Tensor)",
```
Can we look into the semantics here? it seems like device and dtype are to overloads of to that look at different arguments. Do we need to support the device overload as a converter?

narendasan replied:
We could probably lower out device to calls

peri044 replied:
So there are two variants of torch.to usage
1. `torch.to(torch.float32)`
2. `torch.to(torch.device('cuda:0', torch.float32)`
[https://pytorch.org/docs/stable/tensors.html#torch.Tensor.to](url)
Are you suggesting lowering of device overload ?

peri044 replied:
Okay. for some reason your second comment didn't show up when I saw. I'll move it to lowering

peri044 commented:
@narendasan 
- Added lowering of aten::to.device with testcase
- Added support for another variant aten::to.other which casts into datatype of other tensor.
- I should change the doc at this place right ? [https://github.com/NVIDIA/TRTorch/blob/master/docs/_sources/contributors/lowering.rst.txt#remove-to](url)

narendasan commented on core/conversion/converters/impl/constant.cpp:
```diff
@@ -16,13 +16,16 @@ auto constant_registrations TRTORCH_UNUSED = RegisterNodeConversionPatterns()
-              const_layer->setName(util::node_info(n).c_str());
-              auto const_out = ctx->AssociateValueAndTensor(n->outputs()[0], const_layer->getOutput(0));
-
+              nvinfer1::ITensor* const_out;
```
Might need to be rebased

peri044 replied:
done

narendasan commented on core/conversion/evaluators/prim.cpp:
```diff
@@ -101,6 +102,34 @@ auto prim_registrations =
+                    [](const torch::jit::Node* n, kwargs& args) -> c10::optional<torch::jit::IValue> {
+                      auto input = args.at(n->input(0));
+                      if (input.isITensor()) {
+                        auto trt_dtype = input.ITensor()->getType();
```
We have a function TRTDataTypeToScalarType, as well as an optional version which can gracefully tell you there is no type

peri044 replied:
done

peri044 commented on core/conversion/converters/converter_util.cpp:
```diff
@@ -60,6 +60,20 @@ nvinfer1::ITensor* addUnpadding(
   }
 }
 
+bool register_cast_layer(
```
done

narendasan commented on core/conversion/converters/impl/cast.cpp:
```diff
@@ -0,0 +1,82 @@
+             [](ConversionCtx* ctx, const torch::jit::Node* n, args& args) -> bool {
+               auto self = args[0].ITensorOrFreeze(ctx);
+               auto output_dtype = args[1].unwrapToScalar().to<int64_t>();
+               auto aten_to_trt_dtype_map = util::get_aten_trt_type_map();
```
We can just use ScalarToTRTDataType. Use the opt version if you want to handle the error yourself

peri044 replied:
done

narendasan commented on core/conversion/converters/impl/cast.cpp:
```diff
@@ -0,0 +1,82 @@
+             [](ConversionCtx* ctx, const torch::jit::Node* n, args& args) -> bool {
+               auto self = args[0].ITensorOrFreeze(ctx);
+               nvinfer1::DataType other_dtype = args[1].ITensorOrFreeze(ctx)->getType();
+               auto aten_to_trt_dtype_map = util::get_aten_trt_type_map();
```
Same thing here

peri044 replied:
done

narendasan commented on core/conversion/converters/impl/constant.cpp:
```diff
@@ -16,10 +16,16 @@ auto constant_registrations TRTORCH_UNUSED = RegisterNodeConversionPatterns()
-              auto const_out = ctx->AssociateValueAndTensor(n->outputs()[0], tensor_to_const(ctx, t));
-
+              nvinfer1::ITensor* const_out;
+              if (args[0].isITensor()){
```
This should just be tensor_to_const

peri044 replied:
done

narendasan commented on core/lowering/passes/remove_nops.cpp:
```diff
@@ -20,12 +20,10 @@ struct NOPRemoval {
-    LOG_DEBUG(
-        "RemoveNOPs - Note: Removing remaining aten::to operators (in addition to other ops that have no meaning in TRT), if type casts need to be preserved, add a pass before this pass is run");
-    LOG_GRAPH("Post aten::to removal: " << *graph_);
+    LOG_DEBUG("RemoveNOPs - Note: Removing detach operators as they have no meaning in TRT");
```
Just say removing operators that have no meaning in TensorRT

peri044 replied:
done

narendasan commented on core/util/trt_util.h:
```diff
@@ -134,6 +134,7 @@ torch::jit::Node* cloneNode(
     std::shared_ptr<torch::jit::Graph>& graph,
     std::unordered_map<torch::jit::Value*, torch::jit::Value*>& old_to_new);
 const std::unordered_map<at::ScalarType, nvinfer1::DataType>& get_aten_trt_type_map();
+const std::unordered_map<nvinfer1::DataType, at::ScalarType>& get_trt_aten_type_map();
```
We shouldnt need this

peri044 replied:
done

narendasan commented on core/conversion/converters/converter_util.cpp:
```diff
@@ -177,7 +177,11 @@ nvinfer1::ITensor* tensor_to_const(ConversionCtx* ctx, at::Tensor t) {
   LOG_DEBUG(ctx->logger, "Freezing tensor " << tensor_id.str() << " as an IConstantLayer");
-  const_layer->setName(("[Freeze Tensor " + tensor_id.str() + " ]").c_str());
+  if (!name.empty()) {
+    const_layer->setName(name.c_str());
```
Should this set the name of the tensor or the name of the layer? I think its easier if the message `[Freeze Tensor " + name + " ]` is there for debugging, but having an actual name for the tensor might be nice

peri044 replied:
I think it sets the name of the layer. 
Yeah by default, it sets the one you mentioned `[Freeze Tensor " + name + " ]` 
This name argument is optional if anyone wants to set it to a more explicit name. 
For example, when you check the verbose log of TensorRT tactics, most of these constants are named `Freeze Tensor ....`
If anyone wants to find an explicit constant by name, it would be nice to have I thought.

narendasan replied:
I mean the constant layer i think is always freezing a tensor, what i meant is we can have some conditional where its like 

```
if name:
    tensor_id = name
else: 
    tensor_id = int(&tensor)
const_layer.setName([Freeze tensor {}].(tensor_id))
```

narendasan commented on core/conversion/evaluators/prim.cpp:
```diff
@@ -101,6 +102,28 @@ auto prim_registrations =
+                      } else if (input.isIValue()) {
+                        if (input.IValue()->isTensor()) {
+                          auto pyt_input = input.IValue()->toTensor();
+                          return static_cast<int>(c10::typeMetaToScalarType(pyt_input.dtype()));
```
Any reason to not use pyt_input.scalar_type() directly? would they do different things sometimes?

peri044 replied:
done

narendasan commented:
minor stuff, tests pass, once those are resolved ready to merge

narendasan commented on tests/core/conversion/converters/test_cast.cpp:
```diff
@@ -0,0 +1,164 @@
+  auto in1 = at::randint(1, 3, {3, 4, 3}, {at::kCUDA});
+  auto in2 = at::randint(1, 3, {3, 4, 3}, {at::kCUDA});
+  // Lower aten::type_as to aten::to.other
+  trtorch::core::lowering::passes::ReduceToOperation(g);
```
Does this test actually test what we care about? I think that it should be two tensors of different types and then checking the output tensors of the JIT graph are the same type
